server/src/main/java/org/elasticsearch/transport/TransportService.java

```diff
@@ -573,72 +573,64 @@ private <T extends TransportResponse> void sendRequestInternal(final Transport.C
             throw new IllegalStateException("can't send request to a null connection");
         }
         DiscoveryNode node = connection.getNode();
-
-        TimeoutResponseHandler<T> responseHandler = new TimeoutResponseHandler<>(handler);
-        // TODO we can probably fold this entire request ID dance into connection.sendRequest but it will be a bigger refactoring
-        final long requestId = responseHandlers.add(new Transport.ResponseContext<>(responseHandler, connection, action));
+        if (lifecycle.stoppedOrClosed()) {
+            handler.handleException(new SendRequestTransportException(node, action, new NodeClosedException(localNode)));
+            return;
+        }
+        final long requestId = responseHandlers.newRequestId();
         final TimeoutHandler timeoutHandler;
         if (options.timeout() != null) {
-            timeoutHandler = new TimeoutHandler(requestId, connection.getNode(), action);
-            responseHandler.setTimeoutHandler(timeoutHandler);
+            timeoutHandler = new TimeoutHandler(requestId, connection.getNode(), action, options.timeout());
```

Comment:
For really short timeouts I guess this could introduce a race because this now already schedules the timeout logic and the handler isn't registered yet. It gets registered almost right afterwards, so this should only be an issue if there is a timeout in the microseconds range(?).

Could also change it to still trigger the schedule after and just change the variable from `volatile` to a regular mutable field.

Reply:
+1 to point about immediate schedule in constructor, but not sure I got another option - 
does `still trigger the schedule after ` mean restoring `private void scheduleTimeout(TimeValue timeout)`?

Reply:
> does still trigger the schedule after mean restoring private void scheduleTimeout(TimeValue timeout)?

yes it would mean restoring it. (but non-volatile, since it is not needed)

Reply:
> volatile is not needed 
 
 ah, because of 
 https://github.com/crate/crate/pull/12528/files#diff-27dc81a65a8d02c5e6ea7e5a86e91659cb8b48c9aed0faa3fc77cd3c73cfdaebR976 and removal of `setTimeoutHandler`. sounds good

Reply:
Is there any chance, we can make the field final without introducing a potential race condition ?

Reply:
Pushed a fixup